File: bigo/merge_sort.c
#include<stdio.h>

void merge_array(int a[], int left, int mid, int right,int buf[])
{
    int i = left, j = mid+1;
    int m = mid, n = right;
    int k = 0;

    while(i <= m && j <= n){    // <=
        if(a[i] < a[j]){
            buf[k++] = a[i++];
        }else{
            buf[k++] = a[j++];
        }   
    }   
    while(i<=m)
        buf[k++] = a[i++];

    while(j<=n)
        buf[k++] = a[j++];

    for(i=0;i<k;i++)
        a[left + i] = buf[i];   //a[left + 1]

}

void merge_sort(int a[], int left, int right, int buf[])
{
    if(left < right)   //  <
    {   
        int mid = left + ((right-left)>>1);
        merge_sort(a, left, mid, buf);
        merge_sort(a, mid+1, right, buf);

        merge_array(a,left,mid,right,buf);
    }   

}

int main()
{
    int a[] = {12,2,45,-20,65,46,1,413,243,241,2,1,-3};
    int len = sizeof(a)/sizeof(int);
    int buf[30]={0};

    merge_sort(a,0,len-1,buf);

    int i;
    for(i=0;i<len;i++){
        printf("%d\n", a[i]);
    }

}

File: bigo/merge_sort.java
package algorithm;

public class Merge {
	
	static void merge(int [] arr, int left, int mid ,int right){
		
		int [] tmp = new int[arr.length];
		int k=0;
		int i=left, j=mid;
		int m=mid+1, n=right;
		
		while(i<=j && m<=n){
			if(arr[i] < arr[m]){
				tmp[k++] = arr[i++];
			}else{
				tmp[k++] = arr[m++];
			}
		}
		
		while(i<=j){
			tmp[k++] = arr[i++];
		}
		
		while(m<=n){
			tmp[k++] = arr[m++];
		}
		
		for(int x=0;x<k;x++){
			arr[x+left] = tmp[x];
		}
	}
	
	static void mergeSort(int [] tmp,int left, int right){

		if(left < right){
			int mid = left + (right-left)/2;
			
			mergeSort(tmp, left, mid);
			mergeSort(tmp, mid+1, right);
		
			merge(tmp,left,mid,right);
		
		}
		
		
	}
	
	public static void main(String[] args) {
	
		int [] tmp = {7,6,5,4,3,2,1,10,9,8};
		int  length = tmp.length-1;
		
		mergeSort(tmp,0, length);
		
		for (int i=0; i < tmp.length; i++){
			System.out.println(tmp[i]);
		}
	}
}
